lib/storage/dataset-clients.js

```diff
@@ -154,8 +154,10 @@ function upsertOrDeleteDatasetRecords(datasetId, datasetClientId, records, cb) {
     } else if (op === 'delete') {
       //remove the ref
       update['$pull'] = {'refs': datasetClientId};
-      // also set the hash to null to handle race condition when multiple refs associated with single record
-      update['$set']['hash'] = null;
+      if(datasetClientId && datasetClientId.length !==0){
```

Comment:
Actually this is still quite bad. Technically we should be 

A) Removing an item if there is no other ref
B) Removing just ref if there are other refs